src/GafferArnoldUI/BarndoorVisualiser.cpp

```diff
@@ -0,0 +1,290 @@
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "IECore/MeshPrimitive.h"
+#include "IECoreGL/CurvesPrimitive.h"
+#include "IECore/Shader.h"
+#include "IECoreGL/Group.h"
+#include "IECoreGL/ToGLMeshConverter.h"
+#include "IECoreGL/ShaderStateComponent.h"
+#include "IECoreGL/ShaderLoader.h"
+#include "IECoreGL/TextureLoader.h"
+
+#include "Gaffer/Metadata.h"
+
+#include "GafferSceneUI/LightFilterVisualiser.h"
+#include "GafferArnoldUI/BarndoorVisualiser.h"
+
+using namespace Imath;
+using namespace IECore;
+using namespace IECoreGL;
+using namespace GafferArnoldUI;
+
+namespace
+{
+// \todo The following two functions have been stolen from StandardLightVisualiser.
+//       Should be accessible to more users?
+template<typename T>
+T parameter( InternedString metadataTarget, const IECore::CompoundData *parameters, InternedString parameterNameMetadata, T defaultValue )
+{
+	ConstStringDataPtr parameterName = Gaffer::Metadata::value<StringData>( metadataTarget, parameterNameMetadata );
+	if( !parameterName )
+	{
+		return defaultValue;
+	}
+
+	typedef IECore::TypedData<T> DataType;
+	if( const DataType *parameterData = parameters->member<DataType>( parameterName->readable() ) )
+	{
+		return parameterData->readable();
+	}
+
+	return defaultValue;
+}
+
+void spotlightParameters( InternedString metadataTarget, const IECore::CompoundData *parameters, float &innerAngle, float &outerAngle, float &lensRadius )
```

Comment:
I think I mentioned that we might provide some sort of LightAccessor class at some point, to provide an interface for setting and getting parameters like these. We could then use that in a LightSpreadsheet and light manipulators etc. Seems like we have another use case here. Rather than send this work off on a tangent though, perhaps we should just provide something like this function as a static method on the StandardLightVisualiser?

If we do that, I'd like to consider what the most appropriate arguments might be - it seems that `metadataTarget` and `parameters` are specific to the internal implementation of StandardLightVisualiser, and a future LightAccessor abstraction wouldn't use them either. Perhaps the most logical arguments are `lightAttributeName` and `const ObjectVector *lightShader`? Those are the highest-level representation of a light we have in our data flow, and require the least knowledge from the client (hence why they're also the arguments to `LightVisualiser::visualise()?`. Could that work?